File: frontend/lib/features/profile/data/models/requests.dart
class UpdateProfileRequestModel {
  final String locale;
  final String theme;
  final bool hasSeenQuestions;
  final String? ageCategory;
  final String? gender;
  final String? continent;
  final String? country;
  final String? region;
  final String? activity;
  final String? financialSituation;
  final bool? livesInUrbanArea;
  final String? relationshipStatus;
  final String? levelOfEducation;
  final bool? hasChildren;

  final bool notificationsEnabled;
  final bool notificationsForPrivateMessagesEnabled;
  final bool notificationsForPublicMessageLikedEnabled;
  final bool notificationsForPublicMessageRepliesEnabled;
  final bool notificationsUserJoinedYourChallengeEnabled;
  final bool notificationsUserDuplicatedYourChallengeEnabled;

  const UpdateProfileRequestModel({
    required this.locale,
    required this.theme,
    required this.hasSeenQuestions,
    this.ageCategory,
    this.gender,
    this.continent,
    this.country,
    this.region,
    this.activity,
    this.financialSituation,
    this.livesInUrbanArea,
    this.relationshipStatus,
    this.levelOfEducation,
    this.hasChildren,
    required this.notificationsEnabled,
    required this.notificationsForPrivateMessagesEnabled,
    required this.notificationsForPublicMessageLikedEnabled,
    required this.notificationsForPublicMessageRepliesEnabled,
    required this.notificationsUserJoinedYourChallengeEnabled,
    required this.notificationsUserDuplicatedYourChallengeEnabled,
  });

  Map<String, dynamic> toJson() {
    return {
      'locale': locale,
      'theme': theme,
      'has_seen_questions': hasSeenQuestions,
      'age_category': ageCategory,
      'gender': gender,
      'continent': continent,
      'country': country,
      'region': region,
      'activity': activity,
      'financial_situation': financialSituation,
      'lives_in_urban_area': livesInUrbanArea,
      'relationship_status': relationshipStatus,
      'level_of_education': levelOfEducation,
      'has_children': hasChildren,
      'notifications_enabled': notificationsEnabled,
      'notifications_for_private_messages_enabled':
          notificationsForPrivateMessagesEnabled,
      'notifications_for_public_message_liked_enabled':
          notificationsForPublicMessageLikedEnabled,
      'notifications_for_public_message_replies_enabled':
          notificationsForPublicMessageRepliesEnabled,
      'notifications_user_joined_your_challenge_enabled':
          notificationsUserJoinedYourChallengeEnabled,
      'notifications_user_duplicated_your_challenge_enabled':
          notificationsUserDuplicatedYourChallengeEnabled,
    };
  }
}

class SetPasswordRequestModel {
  final String newPassword;

  const SetPasswordRequestModel({required this.newPassword});

  Map<String, dynamic> toJson() {
    return {
      'new_password': newPassword,
    };
  }
}

class UpdatePasswordRequestModel {
  final String currentPassword;
  final String newPassword;

  const UpdatePasswordRequestModel(
      {required this.currentPassword, required this.newPassword});

  Map<String, dynamic> toJson() {
    return {
      'current_password': currentPassword,
      'new_password': newPassword,
    };
  }
}

File: backend/src/features/profile/structs/requests.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Deserialize, Serialize)]
pub struct UserUpdateRequest {
    pub locale: String,
    pub theme: String,
    pub timezone: String,

    pub has_seen_questions: bool,
    pub age_category: Option<String>,
    pub gender: Option<String>,
    pub continent: Option<String>,
    pub country: Option<String>,
    pub region: Option<String>,
    pub activity: Option<String>,
    pub financial_situation: Option<String>,
    pub lives_in_urban_area: Option<bool>,
    pub relationship_status: Option<String>,
    pub level_of_education: Option<String>,
    pub has_children: Option<bool>,

    pub notifications_enabled: bool,
    pub notifications_for_private_messages_enabled: bool,
    pub notifications_for_public_message_liked_enabled: bool,
    pub notifications_for_public_message_replies_enabled: bool,
    pub notifications_user_joined_your_challenge_enabled: bool,
    pub notifications_user_duplicated_your_challenge_enabled: bool,
}

#[derive(Debug, Deserialize)]
pub struct IsOtpEnabledRequest {
    pub username: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct SetUserPasswordRequest {
    pub new_password: String,
    pub private_key_encrypted: String,
    pub salt_used_to_derive_key_from_password: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct UpdateUserPasswordRequest {
    pub current_password: String,
    pub new_password: String,
    pub private_key_encrypted: String,
    pub salt_used_to_derive_key_from_password: String,
}

#[derive(Deserialize)]
pub struct DeleteDeviceParams {
    pub token_id: Uuid,
}

#[derive(Deserialize, Serialize)]
pub struct GetUserPublicDataByIdRequest {
    pub user_ids: Vec<Uuid>,
}

#[derive(Deserialize, Serialize)]
pub struct GetUserPublicDataByUsernameRequest {
    pub username: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct SetFcmTokenRequest {
    pub fcm_token: Option<String>,
}
